### PR TITLE
[BD-8573][BpkCarousel][BpkMobileScrollContainer] Resolve force reflow issues causing by carousel utils and scroll container

### PR DESCRIPTION
**Pull request by yifeng-ruan**
After upgrading bpk, performance issues related to force reflow have emerged, particularly with updates to the carousel component. 
To fix the drop in INP Score caused by reflow, I added IntersectionObserver monitoring and requestAnimationFrame to address the reflow caused by scroll and height/width calculations.
image

Remember to include the following changes:

- [x] Ensure the PR title includes the name of the component you are changing so it's clear in the release notes for consumers of the changes in the version e.g `[KOA-123][BpkButton] Updating the colour`
- [ ] `README.md` (If you have created a new component)
- [ ] Component `README.md`
- [ ] Tests
- [ ] Accessibility tests
    - The following checks were performed:
        - [ ] Ability to navigate using a [keyboard only](https://webaim.org/techniques/keyboard/)
        - [ ] Zoom functionality ([Deque University explanation](https://dequeuniversity.com/checklists/web/text)):
            - [ ] The page SHOULD be functional AND readable when only the text is magnified to 200% of its initial size
            - [ ] Pages must reflow as zoom increases up to 400% so that content continues to be presented in only one column i.e. Content MUST NOT require scrolling in two directions (both vertically and horizontally)
        - [ ] Ability to navigate using a [screen reader only](https://webaim.org/articles/screenreader_testing/)
- [ ] Storybook examples created/updated
- [ ] For breaking changes or deprecating components/properties, migration guides added to the description of the PR. If the guide has large changes, consider creating a new Markdown page inside the component's docs folder and link it here

### CODE REVIEW

File: packages/bpk-component-carousel/src/utils.tsx [hunk 1]
```diff
@@ -25,15 +25,27 @@ export function useScrollToInitialImage(
   initialImageIndex: number,
   imagesRef: MutableRefObject<Array<HTMLElement | null>>,
 ) {
-  useEffect(() => {
-    const element = imagesRef.current[initialImageIndex];
+  const handleIntersecting = (index: number) => {
+    const element = imagesRef.current[index];
     if (element) {
       element.scrollIntoView({
```
Comment (yifeng-ruan) [inline packages/bpk-component-carousel/src/utils.tsx:31]: `element.scrollIntoView` can cause reflow performance issues, even if the image is not in the viewport. Therefore, it is necessary to use IntersectionObserver to monitor the element.

File: packages/bpk-component-mobile-scroll-container/src/BpkMobileScrollContainer.js [hunk 2]
```diff
@@ -140,35 +140,38 @@ class BpkMobileScrollContainer extends Component<Props, State> {
...
-    );
-
-    if (!classNames) {
-      return;
-    }
-
-    this.setState(() => ({
-      scrollIndicatorClassName: classNames.join(' '),
-    }));
+    requestAnimationFrame(() => {
```
Comment (yifeng-ruan) [inline packages/bpk-component-mobile-scroll-container/src/BpkMobileScrollContainer.js:143]: computeScrollIndicatorClassName and computeScrollBarAwareHeight will get the height and width of the element, and frequent calls can lead to reflow. Therefore, move the computations to the last frame to improve performance and avoid reflow.

**Comment by arnaugm** [comment]
hi @yifeng-ruan it looks good, can you please add a test that validates the improvements? thanks!

File: packages/bpk-component-carousel/src/utils.test.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { renderHook } from '@testing-library/react-hooks';
+
+import { useScrollToInitialImage } from './utils';
+
+describe('useScrollToInitialImage', () => {
+  it('should scroll to initial image on mount', () => {
```
Comment (arnaugm) [inline packages/bpk-component-carousel/src/utils.test.tsx:6]: this test will still pass without the code optimisation right, how do we test that the optimisation works?

**Review by arnaugm — APPROVED** [review]

**Pull request merged**